packages/product/src/services/product-module-service.ts

```diff
@@ -405,6 +398,229 @@ export default class ProductModuleService<
     return products
   }
 
+  @InjectTransactionManager(shouldForceTransaction, "baseRepository_")
+  protected async update_(
+    data: ProductTypes.UpdateProductDTO[],
+    @MedusaContext() sharedContext: Context = {}
+  ): Promise<TProduct[]> {
+    return await this.baseRepository_.transaction(
+      async (manager) => {
+        const productIds = data.map(pd => pd.id)
+        const existingProductVariants = await this.productVariantService_.list(
+          { product_id: productIds },
+          {},
+          sharedContext
+        )
+
+        const existingProductVariantsMap = new Map<
+          string,
+          ProductVariant[]
+        >(
+          data.map((productData) => [
+            productData.id,
+            existingProductVariants.filter((variant) => variant.product_id === productData.id)
```

Comment:
todo: create a map of it, also, they are existing if there is an id and they are invalid if the variant.product_id does not correspond to the product id, in that case throw

Reply:
```ts
new Map(existingProductVariants.filter((variant) => {
  if (variant.id && variant.product_id !== productData.id) {
    throw
  }
  return variant.id
}).filter(Boolean).map(variant => [variant.id, variant])
```

Reply:
I'm confused, why are we creating a map within a map?

Reply:
To not have a filter to do later for each loop and have only a complexity of O(N) in total

Reply:
I just implemented it, but the code looks really confusing and messy. I'll remove this check entirely.

Reply:
lets keep it as is for now but when creating 100 products with 50 variants we are looping 5000 times instead of 100 times. let see that later, there is way to make it readable and understandable without having a messi code but I would like you to find it :)

Reply:
Your trust in me to find the solution is inspiring 😄